File: 22. Generate Parentheses.cpp
class Solution {
   public:
    char array[1000];
    vector<string> result;
    void dfs(int index, int target, int left, int right) {
        if (left == target && right == target) {
            array[target * 2] = '\0';
            result.push_back(array);
            return;
        }
        if (left <= target && right <= target && left >= right) {
            array[index] = '(';
            dfs(index + 1, target, left + 1, right);

            array[index] = ')';
            dfs(index + 1, target, left, right + 1);
        }
    }
    vector<string> generateParenthesis(int n) {
        result.clear();
        dfs(0, n, 0, 0);
        return result;
    }
};

File: JavaScript/22. Generate Parentheses.js
/**
 * @param {number} n
 * @return {string[]}
 */
var generateParenthesis = function (n) {
    const res = [];

    dfs(0, 0, "");

    function dfs(left, right, str) {
        if (left > n) return;
        if (right > n) return;

        if (str.length === n * 2) {
            res.push(str);
            return;
        }
        if (left >= right) {
            dfs(left + 1, right, str + "(");
            dfs(left, right + 1, str + ")");
        }
    }

    return res;
};
